### DISCUSSION TITLE
Is allowing `?` in functions names a good idea?

### DISCUSSION THREAD

**Discussion by me2beats** [godotengine/godot-proposals#4698] [Scripting]
For example Ruby has this feature
https://ruby-doc.org/core-2.7.0/Array.html#method-i-all-3F

That is this can be used in functions that return bool.
Now GDScript codestyle suggests using `is_` ..
but I can see the problem with `is` — this is a English grammar sensitive word
I mean if we name the function like `is_all_methods_called()`, shouldn't it be named like  `are_all_methods_called()`?

**Comment by me2beats** [comment]
Another potential "pro" for this is you can easily find functions that return boolean in your code — you just type `?` in the search field.

This is also looks more compact to write `?` instead of `is_`

A con maybe is some people can confuse it and something `?` is for in other languages

**Comment by Calinou** [comment]
This would conflict with a [nullable type hint](https://github.com/godotengine/godot-proposals/issues/162) syntax that may be added in the future (such as `int?` or `Object?`), so I'd prefer not.

**Reply by me2beats** [reply]
It's cool that nullable types proposal is approved, but I don't see how they would conflict.
This is only for functions, and nullable types are for types.
The only thing is sometimes they can be confused when you read the code.

**Comment by wareya** [comment]
This would conflict with the most common syntax languages use for short-circuiting null checks, preventing gdscript from adopting it in the future, e.g.

```
var thing = possibly_null_object?.property
# thing is now null if possibly_null_object is null. otherwise it contains the value of "property"
var thing2 = possibly_null_function_object?(arg1, arg2)
# thing2 is now null if possibly_null_function_object is null. otherwise it contains the result of a function call
```

(Rust uses similar syntax for coalescing error types.)

In general, identifiers should allow the fewest number of lower ascii characters possible, to leave the door open for future syntax changes without breaking compatibility. If you want something like a question mark in a variable name, the ideal would be to pull a question mark from upper unicode and use that instead, and insist that identifiers allow printable upper unicode characters in them.

**Comment by theraot** [comment]
I would have no issue with calling a method `are_all_methods_called`. Although, I see the benefit of having methods that return bool together in auto complete, and easy to search by "is_", so consider `is_every_methods_called`.
